endpoints/openrtb2/auction_test.go

```diff
@@ -36,19 +37,244 @@ import (
+		testCaseFiles, err := getTestFiles(test.directory)
+		assert.NoError(t, err, "Test case %s. Error reading files from directory %s \n", test.description, test.directory)
+
+		for _, file := range testCaseFiles {
+			data, err := ioutil.ReadFile(file)
+			assert.NoError(t, err, "Test case %s. Error reading file %s \n", test.description, file)
+
+			runTestCase(t, data, file)
+		}
+	}
+}
+
+func getTestFiles(dir string) ([]string, error) {
+	var filesToAssert []string
+
+	fileList, err := ioutil.ReadDir(dir)
+	if err != nil {
+		return nil, err
+	}
+
+	// Append the path of every file found in `dir` to the `filesToAssert` array
+	for _, fileInfo := range fileList {
+		filesToAssert = append(filesToAssert, filepath.Join(dir, fileInfo.Name()))
+	}
+
+	return filesToAssert, nil
+}
+
+func runTestCase(t *testing.T, fileData []byte, testFile string) {
+	t.Helper()
+
+	// Retrieve values from JSON file
+	test := parseTestFile(t, fileData, testFile)
+
+	test.Config = generateConfigMaps(test.Config)
+
+	// Run test
+	actualCode, actualBidResponse := doRequest(t, test)
+
+	// Assertions
+	assert.Equal(t, test.ExpectedReturnCode, actualCode, "Test failed. Filename: %s \n", testFile)
+
+	// Either assert bid response or expected error
+	if len(test.ExpectedErrorMessage) > 0 {
+		assert.True(t, strings.HasPrefix(actualBidResponse, test.ExpectedErrorMessage), "Actual: %s \nExpected: %s. Filename: %s \n", actualBidResponse, test.ExpectedErrorMessage, testFile)
+	}
+
+	if len(test.ExpectedBidResponse) > 0 {
+		assertBidResponseEqual(t, testFile, []byte(actualBidResponse), test.ExpectedBidResponse)
+	}
+}
+
+func parseTestFile(t *testing.T, fileData []byte, testFile string) testCase {
+	t.Helper()
+
+	parsedTestData := testCase{}
+	var err, errEm error
+
+	// Get testCase values
+	parsedTestData.BidRequest, _, _, err = jsonparser.Get(fileData, "mockBidRequest")
+	assert.NoError(t, err, "Error jsonparsing root.mockBidRequest from file %s. Desc: %v.", testFile, err)
+
+	// Get testCaseConfig values
+	parsedTestData.Config = &testConfigValues{}
+	var jsonTestConfig json.RawMessage
+
+	jsonTestConfig, _, _, err = jsonparser.Get(fileData, "config")
+	if err == nil {
+		err = json.Unmarshal(jsonTestConfig, parsedTestData.Config)
+		assert.NoError(t, err, "Error unmarshaling root.config from file %s. Desc: %v.", testFile, err)
+	}
+
+	// Get the return code we expect PBS to throw back given test's bidRequest and config
+	parsedReturnCode, err := jsonparser.GetInt(fileData, "expectedReturnCode")
+	assert.NoError(t, err, "Error jsonparsing root.code from file %s. Desc: %v.", testFile, err)
+
+	// Get both bid response and error message, if any
+	parsedTestData.ExpectedBidResponse, _, _, err = jsonparser.Get(fileData, "expectedBidResponse")
+	parsedTestData.ExpectedErrorMessage, errEm = jsonparser.GetString(fileData, "expectedErrorMessage")
+
+	// Assert no error in parsing one or the other
+	if parsedReturnCode >= 200 && parsedReturnCode < 300 {
```

Comment:
I don't think we need this condition anymore with the changes made in `runTestCase`. Maybe add an error check to ensure one of them exists and was retrieved without caring about which one?

Reply:
I tried:
```
if err != nil && errEm != nil {
    t.Fatalf("One or the other should be an error, but not both")
}
```
But if `parsedReturnCode` is between 200 and 300, then `err` is the one that snouldn't be an error, and the other way around. I fear that the condition above could lead us to a couple of false non-error scenarios, unless I'm wrong. I think an improvement, could be the statement below, but do you think it's that much readable? The main flaw I find is the ambiguity of the `t.Fatalf` message.
```
statusOk := parsedReturnCode >= 200 && parsedReturnCode < 300
if (statusOk && err != nil) || (!statusOk && errEm != nil) {
    t.Fatalf("Test case error")
}
```
What do you think?

Reply:
Enforcing the existence of an expectation based on a response code seems like we're building a framework that is too opinionated - which could lead to its own issues down the line. I would check that both err and errEm are either nil or KeyPathNotFoundError (from the jsonparser package).

Reply:
Good point. I rewrote the conditional to not rely on `parsedReturnCode` anymore.